controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,442 @@
+
+func (r *MedusaTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("MedusaTask", req.NamespacedName)
+
+	logger.Info("Starting reconciliation for MedusaTask")
+
+	// Fetch the MedusaTask instance
+	instance := &medusav1alpha1.MedusaTask{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		logger.Error(err, "Failed to get MedusaTask")
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	task := instance.DeepCopy()
+
+	// If the task is already finished, there is nothing to do.
+	if taskFinished(task) {
+		logger.Info("Task is already finished", "MedusaTask", req.NamespacedName, "Operation", task.Spec.Operation)
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// First check to see if the task is already in progress
+	if !task.Status.StartTime.IsZero() {
+		// If there is anything in progress, simply requeue the request
+		if len(task.Status.InProgress) > 0 {
+			logger.Info("Tasks already in progress")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+		}
+
+		logger.Info("task complete")
+
+		// Set the finish time
+		// Note that the time here is not accurate, but that is ok. For now we are just
+		// using it as a completion marker.
+		patch := client.MergeFrom(task.DeepCopy())
+		task.Status.FinishTime = metav1.Now()
+		if err := r.Status().Patch(ctx, task, patch); err != nil {
+			logger.Error(err, "failed to patch status with finish time")
+			return ctrl.Result{}, err
+		}
+
+		// Schedule a sync if the task is a purge
+		if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+			r.scheduleSyncForPurge(task)
+		}
+
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// Task hasn't started yet
+	cassdcKey := types.NamespacedName{Namespace: task.Namespace, Name: task.Spec.CassandraDatacenter}
+	cassdc := &cassdcapi.CassandraDatacenter{}
+	err = r.Get(ctx, cassdcKey, cassdc)
+	if err != nil {
+		logger.Error(err, "failed to get cassandradatacenter", "CassandraDatacenter", cassdcKey)
+		return ctrl.Result{}, err
+	}
+
+	pods, err := r.getCassandraDatacenterPods(ctx, cassdc, logger)
+	if err != nil {
+		logger.Error(err, "Failed to get datacenter pods")
+		return ctrl.Result{}, err
+	}
+
+	if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+		return r.purgeOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypeSync {
+		return r.syncOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypePrepareRestore {
+		return r.prepareRestoreOperation(ctx, task, pods, logger)
+	} else {
+		return ctrl.Result{}, fmt.Errorf("unsupported operation %s", task.Spec.Operation)
+	}
+}
+
+func (r *MedusaTaskReconciler) getCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.CassandraDatacenter, logger logr.Logger) ([]corev1.Pod, error) {
```

Comment:
Looks like this is identical to the  `getCassandraDatacenterPods` in medusabackupjob_controller.go. Can we refactor this into a common, shared method or func?

Reply:
sounds good